### PR TITLE
JAMES-2530 WIP [FILTERS] Complete POJOs, add tests and first draft of DTO

### PR DESCRIPTION
**Pull request by blackheaven**

### CODE REVIEW

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java [hunk 1]
```diff
@@ -19,48 +19,313 @@
...
+            private Comparator(String comparator) {
+                this.comparatorName = comparator;
+            }
+            
+            public String asString() {
+                return comparatorName;
+            }
+        }
+
+        public static Condition of(String field, String comparator, String value) {
```
Comment (mbaechler) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java]: If it's for human being to call, having 3 strings as argument looks like a bad idea

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java [hunk 2]
```diff
@@ -19,48 +19,313 @@
...
+                .add("comparator", comparator)
+                .add("value", value)
+                .toString();
+        }
+    }
+
+
+    public static class Action {
+
+        public static Action ofMailboxIds(List<String> mailboxIds) {
```
Comment (mbaechler) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java]: if you don't want to enforce ImmutableList on input, you should copy it before keeping a reference to it

File: server/data/data-jmap/src/test/java/org/apache/james/jmap/api/filtering/FilteringManagementContract.java [hunk 3]
```diff
@@ -30,6 +32,14 @@
 
 public interface FilteringManagementContract {
 
+    public static final String NAME = "a name";
+    public static final Rule.Condition CONDITION = Rule.Condition.of("cc", "contains", "something");
+    public static final Rule.Action ACTION = Rule.Action.ofMailboxIds(Arrays.asList("id-01"));
+    public static final Rule.Builder RULE_BUILER = Rule.builder().name(NAME).condition(CONDITION).action(ACTION);
```
Comment (mbaechler) [inline server/data/data-jmap/src/test/java/org/apache/james/jmap/api/filtering/FilteringManagementContract.java]: builder is not immutable, you should not keep a static instance of it

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java (hunk 1 above)
Comment (blackheaven) [reply server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java]: I've hesitate to do a `Rule.Condition.Builder`. Should I do it?

Comment (mbaechler) [reply server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java]: In this case, I would take Field, Comparator and String as input type.

File: server/data/data-jmap-cassandra/pom.xml [hunk 4]
```diff
@@ -77,6 +87,16 @@
             <artifactId>awaitility</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.junit.vintage</groupId>
+            <artifactId>junit-vintage-engine</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
```
Comment (chibenwa) [inline server/data/data-jmap-cassandra/pom.xml]: You miss the plateforme?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/RulesIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+            .statusCode(200)
+            .body(SECOND_NAME, equalTo("messages"))
+            .body(SECOND_ARGUMENTS + ".list", hasSize(1))
+            .body(SECOND_ARGUMENTS + ".list[0].textBody", equalTo(expectedTextBody))
+            .body(SECOND_ARGUMENTS + ".list[0].from.email", equalTo(expectedFrom))
+            .body(SECOND_ARGUMENTS + ".list[0].to.email", hasSize(1))
+            .body(SECOND_ARGUMENTS + ".list[0].to.email[0]", equalTo(expectedTo));
+    }
+
+    private String getOutboxId(AccessToken accessToken) {
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/RulesIntegrationTest.java]: This seems like something defined in `JmapCommonRequests` class?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/RulesIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+
+        // Then
+        // User 1 should well receive this mail in the filtered mailbox
+        calmlyAwait.atMost(30, TimeUnit.SECONDS)
+            .until(() -> isTextMessageReceived(user1AccessToken, getFilteredMailboxId(user1AccessToken), ORIGINAL_MESSAGE_TEXT_BODY, USER_2, USER_1));
+    }
+
+    @Test
+    public void jmapRuleShouldFilterAMatchingMailWhenSetEmptyThenEffectively() throws Exception {
+        /* Test scenario :
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/RulesIntegrationTest.java]: This feels like tests for GITLAB-734

In a first time we should simply exercise the CRUD (get/set methods)

**Comment by chibenwa** [comment]
See https://github.com/linagora/james-project/pull/1694

**Pull request closed without merging**